contracts/strategy/RWAStrategy.sol

```diff
@@ -0,0 +1,154 @@
+        if (!TokenizedStrategy.isShutdown()) {
+            // deposit any loose funds
+            uint256 balance = asset.balanceOf(address(this));
+            if (balance > minAmount) {
+                uint256 _amount = Math.min(balance, availableDepositLimit(address(this)));
+                asset.transfer(managerAddress, _amount);
+                totalInvested += _amount;
+            }
+        }
+        _totalAssets = totalInvested + asset.balanceOf(address(this));
+    }
+
+    function availableDepositLimit(address /*_owner*/) public view override returns (uint256) {
+        uint256 balance = asset.balanceOf(address(this));
+        // Return the remaining room.
+        return depositLimit < balance ? 0 : depositLimit - balance;
+    }
+
+    function availableWithdrawLimit(address /*_owner*/) public view override returns (uint256) {
+        return TokenizedStrategy.totalIdle() + totalInvested;
+    }
+
+    function _deployFunds(uint256 _amount) internal override {
+        if (_amount > minAmount) {
+            asset.transfer(managerAddress, _amount);
+            totalInvested += _amount;
+        }
+    }
+
+    function _freeFunds(uint256 _amount) internal override {
+        // We don't check available liquidity because we need the tx to
+        // revert if there is not enough liquidity so we don't improperly
+        // pass a loss on to the user withdrawing.
+        asset.safeTransferFrom(managerAddress, address(this), _amount);
+        totalInvested -= _amount;
+    }
+
+    /**
+     * @dev Optional function for a strategist to override that will
+     * allow management to manually withdraw deployed funds from the
+     * yield source if a strategy is shutdown.
+     *
+     * This should attempt to free `_amount`, noting that `_amount` may
+     * be more than is currently deployed.
+     *
+     * NOTE: This will not realize any profits or losses. A separate
+     * {report} will be needed in order to record any profit/loss. If
+     * a report may need to be called after a shutdown it is important
+     * to check if the strategy is shutdown during {_harvestAndReport}
+     * so that it does not simply re-deploy all funds that had been freed.
+     *
+     * EX:
+     *   if(freeAsset > 0 && !TokenizedStrategy.isShutdown()) {
+     *       depositFunds...
+     *    }
+     *
+     * @param _amount The amount of asset to attempt to free.
+     */
+    function _emergencyWithdraw(uint256 _amount) internal override {
+        uint256 amountToWithdraw = _amount;
+        uint256 balance = asset.balanceOf(address(this));
+        if (_amount > balance) {
+            amountToWithdraw = Math.min(_amount - balance, totalInvested);
+            asset.safeTransferFrom(managerAddress, address(this), amountToWithdraw);
+        }
+        asset.transfer(TokenizedStrategy.management(), _amount);
+        totalInvested = totalInvested > amountToWithdraw ? totalInvested - amountToWithdraw : 0;
```

Comment:
this version still looks wrong, in case` _amount < asset.balanceOf(address(this))` we will reduce `totalInvested` without transferring invested balance. for example :
balance = 10,  totalInvested = 10, _amount = 5
```
 function _emergencyWithdraw(uint256 _amount) internal override {
        uint256 amountToWithdraw = _amount; // 5
        uint256 balance = asset.balanceOf(address(this)); // 10
        if (_amount > balance) { // false
            amountToWithdraw = Math.min(_amount - balance, totalInvested);
            asset.safeTransferFrom(managerAddress, address(this), amountToWithdraw);
        }
        asset.transfer(TokenizedStrategy.management(), _amount); // transfer 5, now balance is 5
        totalInvested = totalInvested > amountToWithdraw ? totalInvested - amountToWithdraw : 0; // totalInvested = 5 now
```

Reply:
and if `_amount > balance + totalInvested` it still will try to transfer a full amount and revert

Reply:
I would suggest something like this (it's an untested draft, don't just copy-paste):
```
 function _emergencyWithdraw(uint256 _amount) internal override {
        uint256 balance = asset.balanceOf(address(this));
        if (balance >= _amount) {
            asset.safeTransfer(TokenizedStrategy.management(), _amount);
        } else {
            uint256 amountToWithdraw = Math.min(_amount - balance, totalInvested);
            totalInvested = totalInvested > amountToWithdraw ? totalInvested - amountToWithdraw : 0;
            address recipient = TokenizedStrategy.management();
            asset.safeTransferFrom(managerAddress, recipient, amountToWithdraw);
            asset.transfer(recipient, balance);
        }
    }
```

Reply:
Check my last implementation at [`806b6f1` (#30)](https://github.com/Into-the-Fathom/fathom-vaults-smart-contracts/pull/30/commits/806b6f157484438ba3f298c80cf0f3fc471acee7)